### PR TITLE
xonsh: 0.15.1 -> 0.17.0

### LINKED ISSUE #206062: xonsh: captured subproc is not capturing
### Describe the bug
- Upstream issue: https://github.com/xonsh/xonsh/issues/5003
Since the issue seems to be specific to NixOS, let's track it here.

### Steps To Reproduce
```xonsh
$ !(echo hi) == 'hi\\\\n'
hi
False

$ !(echo hi).out == 'hi\\\\n'
hi
False
```

### Expected behavior
The two commands above should only print "True".

### Additional context
Tested on xonsh 0.13.3 and 0.13.4

### Notify maintainers
@spwhitt @vrthra 

### Metadata
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.82, NixOS, 23.05 (Stoat), 23.05.20221213.1710ed1`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.12.0`
 - nixpkgs: `/nix/store/dk1bqlz87dzsd8g4fq92pscgd60ji2vr-source`

### LINKED ISSUE #308868: xonsh: wrong version of prompt-toolkit library
Xonsh [currently requires](https://github.com/xonsh/xonsh/blob/c3cd47e1da46c94e8678a389d5c758311cfb8f24/pyproject.toml#L86) `prompt-toolkit>=3.0.29,<3.0.41` because of https://github.com/prompt-toolkit/python-prompt-toolkit/issues/1853

However, in unstable the [xonsh package](https://github.com/NixOS/nixpkgs/blob/63c3a29ca82437c87573e4c6919b09a24ea61b0f/pkgs/by-name/xo/xonsh/unwrapped.nix#L34) is using [v3.0.43](https://github.com/NixOS/nixpkgs/blob/63c3a29ca82437c87573e4c6919b09a24ea61b0f/pkgs/development/python-modules/prompt-toolkit/default.nix#L12)

### PR DESCRIPTION
**Pull request by SamLukeYes**
## Description of changes

Adopted from https://github.com/NixOS/nixpkgs/pull/305316

Fixes https://github.com/NixOS/nixpkgs/issues/308868
Fixes https://github.com/NixOS/nixpkgs/issues/206062

Closes https://github.com/NixOS/nixpkgs/pull/301670

Upstream changes: https://github.com/xonsh/xonsh/compare/0.15.1...0.17.0

Notably, `~/.xonshrc` is no longer loaded by default in non-interactive mode. See https://github.com/xonsh/xonsh/pull/5491

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/xo/xonsh/package.nix [hunk 1]
```diff
@@ -1,23 +1,114 @@
...
+
+    disabledTests = [
+      # fails on sandbox
+      "test_colorize_file"
+      "test_loading_correctly"
+      "test_no_command_path_completion"
+      "test_bsd_man_page_completions"
+      "test_xonsh_activator"
+      "test_subproc_output_format"
+      "test_raise_subproc_error_with_show_traceback"
```
Comment (SamLukeYes) [inline pkgs/by-name/xo/xonsh/package.nix:59]: These are 2 new tests that are failing. Not sure if they should be added here or under "fails on non-interactive shells", or be reported to upstream.

File: pkgs/by-name/xo/xonsh/package.nix [hunk 2]
```diff
@@ -1,23 +1,114 @@
...
 let
-  xonsh-unwrapped = callPackage ./unwrapped.nix { };
-  inherit (xonsh-unwrapped.passthru) python;
 
-  pythonEnv = python.withPackages (ps: [
-    (ps.toPythonModule xonsh-unwrapped)
-  ] ++ extraPackages ps);
+  argset = {
+    pname = "xonsh";
+    version = "unstable-2024-05-25";
```
Comment (Aleksanaa) [inline pkgs/by-name/xo/xonsh/package.nix:16]:
```suggestion
    version = "0.16.0-unstable-2024-05-25";
```

File: pkgs/by-name/xo/xonsh/package.nix (hunk 1 above)
Comment (SamLukeYes) [reply pkgs/by-name/xo/xonsh/package.nix]: `test_subproc_output_format` is marked as a flaky test. Disabling all the flaky tests now.

**Review by LordGrimmauld — COMMENTED** [review]
Maybe also put the compare link in the PR message:
https://github.com/xonsh/xonsh/compare/0.15.1...1613914d507884487251ecafb55e281699931eb8

File: pkgs/top-level/python-packages.nix [hunk 3]
```diff
@@ -17129,6 +17129,8 @@ self: super: with self; {
 
   xnd = callPackage ../development/python-modules/xnd { };
 
+  xonsh = toPythonModule (pkgs.xonsh.override { });
```
[the remark this answers is not in the document]
Comment (Aleksanaa) [reply pkgs/top-level/python-packages.nix:17132]: Here should  be `xonsh = toPythonModule (pkgs.xonsh.override { python3Packages = self; });`? Not sure

**Review by LordGrimmauld — APPROVED** [review]

**Comment by drmikecrowe** [comment]
Can this move to `0.16.0`  instead of `0.16.0-unstable-2024-05-25`?

**Comment by AndersonTorres** [comment]
> Can this move to `0.16.0` instead of `0.16.0-unstable-2024-05-25`?

And using an outdated prompt-toolkit in a wild override stack that will be removed in the next release? Nope.

File: pkgs/top-level/aliases.nix [hunk 4]
```diff
@@ -1398,7 +1398,7 @@ mapAliases ({
   xineLib = xine-lib; # Added 2021-04-27
   xineUI = xine-ui; # Added 2021-04-27
   xmlada = gnatPackages.xmlada; # Added 2024-02-25
-  xonsh-unwrapped = xonsh.passthru.unwrapped;
+  xonsh-unwrapped = python3Packages.xonsh;
```
Comment (AndersonTorres) [inline pkgs/top-level/aliases.nix:1401]: Needs `# Added <date>`

**Review by AndersonTorres — APPROVED** [review]

**Comment by SamLukeYes** [comment]
0.17.0 was released just now. Let's use this stable version instead.

**Review by AndersonTorres — APPROVED** [review]

**Comment by cvoges12** [comment]
LGTM

**Comment by cvoges12** [comment]
Can we backport this to 24.05?

**Comment by SamLukeYes** [comment]
> Can we backport this to 24.05?

I'm not so sure about it tho, as it introduces at least one breaking change (https://github.com/xonsh/xonsh/pull/5491)

**Pull request merged**